### ISSUE TITLE
firestore: Updates in transactions fail silently

### ISSUE THREAD

**Issue by lc-jordanfelix** [googleapis/google-cloud-go#1529]
## Client

Firestore - 	"cloud.google.com/go/firestore" cloud.google.com/go v0.43.0

## Describe Your Environment

Google Cloud Run

## Expected Behavior

Successful committed transactions with tx.Update/tx.Set usages result in updates to a document

## Actual Behavior

Frequently, but not always, the document will have no change.  There is no error returned from any of the tx function calls or transaction commit.  All code within the transaction is executed successfully, yet when we inspect the document in firestore, we can see that the transaction was not applied.

We have tried using tx.Update and tx.Set and both of these have the same intermittent issue.  

We have followed the [documentation](https://cloud.google.com/firestore/docs/manage-data/transactions#transactions) for the go code, with minor tweaks. See the snippet below:

```go
documentRef := c.engageFirebaseClients.Store.Collection("messages_v2").Doc(messageChannelSentUpdate.MessageChannelId)
if err := c.engageFirebaseClients.Store.RunTransaction(ctx, func(ctx context.Context, tx *firestore.Transaction) error {

	document, err := tx.Get(documentRef)
	if err != nil {
		return lib_errors.Wrap(err, "Failed getting document from documentRef")
	}

	documentBytes, err := json.Marshal(document.Data())
	if err != nil {
		return lib_errors.Wrap(err, "Failed json marshalling document data")
	}
	var messageChannelMessage messageChannelMessage
	if err := json.Unmarshal(documentBytes, &messageChannelMessage); err != nil {
		return lib_errors.Wrap(err, "Failed json unmarshalling documentBytes to messageChannelMessage")
	}

	updateMap := generateMessageV2FirestoreMessageUpdates(ctx, messageChannelSentUpdate, messageChannelMessage)
	if len(updateMap) > 0 {
		if err := tx.Set(documentRef, updateMap, firestore.MergeAll); err != nil {
			return lib_errors.Wrap(err, "Failed setting transaction updates")
		}
	}

	return nil

}); err != nil {
	return lib_errors.Wrap(err, "Failed running transaction")
}
```

In regards to  `if len(updateMap) > 0 ` we have logs of the updateMap where we can see that it has updates during requests where the transaction fails silently as we log our updateMap before returning in the `generateMessageV2FirestoreMessageUpdates` function.

We have run many tests in a row with the same data but different documents.  It succeeds sometimes, but fails quite often.

**Comment by jadekler** [comment]
@schmidt-sebastian is this related to https://github.com/googleapis/google-cloud-go/issues/1412, do you think?

**Comment by schmidt-sebastian** [comment]
This might be related. We do need to "bubble up" these user errors. 

@lc-jordanfelix Instead of returning errors, can you log your errors? It would be interesting to see if one of your error cases applies here. We might just be swallowing the error message and not committing the changes.

**Comment by lc-jordanfelix** [comment]
We do log all of our errors further up the chain, even when "treating as success," and there are no error logs from these failures.

**Comment by lc-jordanfelix** [comment]
We have added the logs and there are no errors occurring.

**Comment by jadekler** [comment]
I'm having a hard time in the code where we might be receiving an error but not returning it to the user. @lc-jordanfelix Could you provide us a minimal reproduction (perhaps an integration test that shows this when ran with `go test . -v -run 1000` or something)?

**Comment by lc-jordanfelix** [comment]
@jadekler Sorry, I believe we've managed to solve it after a bunch of testing.  Thanks for taking the time to look into it, it helped us solve it as after writing a light weight go tool to run a create/update/read a couple thousand times, we found it was only happening during our integration test.

We were creating these Firestore documents through a queue and sometimes we were getting duplicate messages.  This ended up being the issue due to us using `Set()` to create each document, so when we did an update to a document and then the duplicate message was processed after the update, it would overwrite the update with the "new" `Set()` data.  Then when we read the document, there was no indication that the document was ever updated as the `Set()` would set the document back to it's original state.

We now are using `Create()` and then treating conflicts as a success.
